docs/2019-05-23-Unification-Modulo-Equations.md

```diff
@@ -0,0 +1,163 @@
+`concat` - concatenates two lists.
+Heating and cooling could be another example.
+
+Proposed Solution
+-----------------
+
+The first attempt to solve this would be to define some sort of
+no-junk/no-confusion axiom
+for that sort (e.g. any list can be built with `unit`, `elem` and `concat`),
+and then expand the two terms in all possible ways using the available
+equations, then match the two terms at the top as long as we stay within
+the same sort, then use unification for the remainders.
+
+However, this has a few problems. First, there is no way of proving that
+this is the most general unifier, i.e. there is no way of showing that we
+truly expanded the term in all possible ways. Second, this would be really
+inefficient.
+
+### Actual proposal
+
+The second attempt, however, would be to define some unification axioms (I
+think that they are, indeed, axioms, but they could be theorems if we find
+a way to infer them), specifying how to solve certain unification cases
+in a way that can be used almost directly by the Haskell backend (some
+changes needed).
+
+These axioms would allow us to reduce the hard to handle
+unification cases to unambiguous ones (e.g. reduce AC unification to
+constructor-based unification).
+
+See the use-cases below for examples and explanations.
+
+We could write these in a way that we don't duplicate unification cases.
+Then we could apply these as simplification axioms, defining our unification
+strategy in k. When we need performance, we could implement the ones that are
+used more frequently in Haskell.
+
+Or, if that ever becomes desirable (e.g. if we want it to make proof
+certificates easier to generate), we could give up on performance and define
+all simplification cases as axioms/theorems.
+
+Also, we should probably use set variables, to avoid the large number of
+checks needed when working with normal variables (though, of course, we
+need to make sure that the axioms still hold).
+
+Use-case: lists
+---------------
+
+For lists, we could have the following axioms
+```
+⌈elem(x)∧unit⌉ = ⊥
+⌈elem(x)∧elem(y)⌉ = ⌈x∧y⌉
+⌈concat(x,y)∧unit⌉ = ⌈x∧unit⌉ ∧ ⌈y∧unit⌉
+⌈concat(elem(x),y) ∧ concat(elem(z),t)⌉ = ⌈x∧z⌉ ∧ ⌈y∧t⌉
+```
+This is missing some axioms (e.g. the ones unifying at the end of the list,
+or splitting the list), but it looks like these, together with the usual list
+axioms and the `and` commutativity axioms could be enough to solve list
+unification.
+
+Note that applying these axioms allows us to reduce complex expressions to
+simpler ones, without worrying about how many terms are equivalent to the
+one we have: we just have to bring the term in one of the forms required for
+one of these axioms, then we apply the axiom and that’s it.
+
+As an example, say we need to unify `concat(concat(elem(φ_1), elem(φ_2)), φ_3)`
+with `concat(concat(unit, elem(ψ_1)), concat(elem(ψ_2), ψ_3))`. We first use
+the list axioms to rewrite the first term as
```

Comment:
The "list axioms" here, I guess, refer to the assoc axiom of `concat`. Such an axiom can be applied in both directions. How do we know which direction should we use? In general, the approach here looks like the one that uses theory axioms (e.g., assoc, comm, etc.) to transform terms to some sort of canonical forms, and then use unification axioms (after orientation) as rewriting rules to solve the unification problem. Indeed, the unification rules seem to require that both terms have `concat` associative to the right. 
If my above understanding is correct, then the problem is that (almost) all theory axioms are not simplification rules, and they do not yield canonical forms in general because they can be applied in both directions, and no one direction is more preferable than the other.

Reply:
I hope the answer above covers this. Please tell me if it doesn't.